### PR TITLE
feat(data-gen): support KAFKA format in DataGen

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

ksql-data-gen now supports a `key-format` and `value-format`, rather than the previous `format`.
`format` is still accepted to maintain backwards compatibility, however, its use is deprecated.

`key-format` supports the existing formats and the new `KAFKA` format that writes data using Kafka's inbuilt serializers.
`value-format` won't work with the `KAFKA` format as the value schemas have multiple fields.

`key-format` defaults to `KAFKA`. This is backwards compatible as keys were previously serialized using `StringSerializer`.
`value-format` defaults to `JSON`, as was the case for the old `format` option.

As part of this work I've moved some functionality from the `IntegrationTestHarness` which isn't available in `ksql-examples` into our test Kafka cluster class, which is available.

### Testing done 

Functional test added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/tutorials/generate-custom-test-data.rst [hunk 1]
```diff
@@ -47,7 +47,8 @@ Required Arguments
 Name                                        Default  Description
 ==========================================  =======  ===========================================================================================
 ``schema=<avro schema file>``                        Path to an Avro schema file. Requires the ``format``, ``topic``, and ``key`` options.
-``format=<record format>``                    json   Format of generated records: one of ``avro``, ``json``, or ``delimited``. Case-insensitive.
+``key-format=<key format>``                    kafka Format of generated record keys: one of ``avro``, ``json``, ``delimited``, ``kafka``. Case-insensitive.
```
Comment (JimGalasyn) [inline docs/tutorials/generate-custom-test-data.rst]:
```suggestion
``key-format=<key format>``                    Kafka format of generated record keys: one of ``avro``, ``json``, ``delimited``, ``kafka``. Case-insensitive.
```

File: docs/tutorials/generate-custom-test-data.rst [hunk 2]
```diff
@@ -47,7 +47,8 @@ Required Arguments
 Name                                        Default  Description
 ==========================================  =======  ===========================================================================================
 ``schema=<avro schema file>``                        Path to an Avro schema file. Requires the ``format``, ``topic``, and ``key`` options.
-``format=<record format>``                    json   Format of generated records: one of ``avro``, ``json``, or ``delimited``. Case-insensitive.
+``key-format=<key format>``                    kafka Format of generated record keys: one of ``avro``, ``json``, ``delimited``, ``kafka``. Case-insensitive.
+``value-format=<value format>``                json  Format of generated record values: one of ``avro``, ``json``, ``delimited``. Case-insensitive.
```
Comment (JimGalasyn) [inline docs/tutorials/generate-custom-test-data.rst]:
```suggestion
``value-format=<value format>``                JSON format of generated record values: one of ``avro``, ``json``, ``delimited``. Case-insensitive.
```

File: docs/tutorials/generate-custom-test-data.rst [hunk 3]
```diff
@@ -70,7 +71,8 @@ The following options apply to both the ``schema`` and ``quickstart`` options.
 Name                                          Default                                              Description
 ============================================  ===================================================  =========================================================================================
 ``bootstrap-server=<kafka-server>:<port>``    localhost:9092                                       IP address and port for the Kafka server to connect to.
-``format=<record format>``                    json                                                 Format of generated records: ``avro``, ``json``, or ``delimited``. Case-insensitive. Required by the ``schema`` option.
+``key-format=<key format>``                   kafka                                                Format of generated record keys: ``avro``, ``json``, ``delimited`` or ``kafka``. Case-insensitive. Required by the ``schema`` option.
```
Comment (JimGalasyn) [inline docs/tutorials/generate-custom-test-data.rst]:
```suggestion
``key-format=<key format>``                   Kafka                                                format of generated record keys: ``avro``, ``json``, ``delimited`` or ``kafka``. Case-insensitive. Required by the ``schema`` option.
```

File: docs/tutorials/generate-custom-test-data.rst [hunk 4]
```diff
@@ -70,7 +71,8 @@ The following options apply to both the ``schema`` and ``quickstart`` options.
 Name                                          Default                                              Description
 ============================================  ===================================================  =========================================================================================
 ``bootstrap-server=<kafka-server>:<port>``    localhost:9092                                       IP address and port for the Kafka server to connect to.
-``format=<record format>``                    json                                                 Format of generated records: ``avro``, ``json``, or ``delimited``. Case-insensitive. Required by the ``schema`` option.
+``key-format=<key format>``                   kafka                                                Format of generated record keys: ``avro``, ``json``, ``delimited`` or ``kafka``. Case-insensitive. Required by the ``schema`` option.
+``value-format=<value format>``               json                                                 Format of generated record values: ``avro``, ``json``, or ``delimited``. Case-insensitive. Required by the ``schema`` option.
```
Comment (JimGalasyn) [inline docs/tutorials/generate-custom-test-data.rst]:
```suggestion
``value-format=<value format>``               JSON                                                 format of generated record values: ``avro``, ``json``, or ``delimited``. Case-insensitive. Required by the ``schema`` option.
```

File: ksql-examples/src/main/java/io/confluent/ksql/datagen/ProducerFactory.java [hunk 5]
```diff
@@ -15,26 +15,124 @@
...
+    final Optional<SchemaRegistryClient> srClient = SchemaRegistryClientFactory
+        .getSrClient(keyFormat, valueFormat, props);
+
+    return new DataGenProducer(keyFactory, valueFactory, srClient::get);
+  }
+
+  private static KsqlSerdeFactory createSerdeFactory(final Format format) {
+    final KsqlSerdeFactory structSerdeFactory = SERDE_FACTORIES.create(format, Optional.empty());
+
+    if (format.supportsUnwrapping()) {
```
Comment (agavra) [inline ksql-examples/src/main/java/io/confluent/ksql/datagen/ProducerFactory.java]: did you mean `!format.supportsUnwrapping()`? are there unit tests for this?

File: ksql-examples/src/main/java/io/confluent/ksql/datagen/SchemaRegistryClientFactory.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    if (keyFormat != Format.AVRO && valueFormat != Format.AVRO) {
+      return Optional.empty();
+    }
+
+    final KsqlConfig ksqlConfig = new KsqlConfig(props);
+    if (ksqlConfig.getString(KsqlConfig.SCHEMA_REGISTRY_URL_PROPERTY) == null) {
+      throw new KsqlException("Schema registry url is not set.");
+    }
+
+    return Optional.of(new CachedSchemaRegistryClient(
```
Comment (agavra) [inline ksql-examples/src/main/java/io/confluent/ksql/datagen/SchemaRegistryClientFactory.java]: I remember seeing this passed in as a method reference - isn't `CachedSchemaRegistryClient` thread safe? why create a new one each time?

File: ksql-examples/src/main/java/io/confluent/ksql/datagen/DataGenProducer.java [hunk 7]
```diff
@@ -96,6 +131,31 @@ public void populateTopic(
     producer.close();
   }
 
+  @SuppressWarnings("unchecked")
+  private Serializer<String> getKeySerializer(final KsqlConfig ksqConfig) {
+    final PersistenceSchema schema = keySerializerFactory.getFormat().supportsUnwrapping()
+        ? PersistenceSchema.of((ConnectSchema) KEY_SCHEMA.fields().get(0).schema())
+        : PersistenceSchema.of(KEY_SCHEMA);
+
+    return (Serializer<String>) (Serializer)keySerializerFactory
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-examples/src/main/java/io/confluent/ksql/datagen/DataGenProducer.java]: That won't compile... at least not yet. The generics are all messed up for now.

File: ksql-examples/src/main/java/io/confluent/ksql/datagen/SchemaRegistryClientFactory.java (hunk 6 above)
Comment (big-andy-coates) [reply ksql-examples/src/main/java/io/confluent/ksql/datagen/SchemaRegistryClientFactory.java]: `CachedSchemaRegistryClient` is not thread safe!

File: ksql-examples/src/main/java/io/confluent/ksql/datagen/ProducerFactory.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-examples/src/main/java/io/confluent/ksql/datagen/ProducerFactory.java]: Nope, it's just the name of `UnwrappedSerdeFactory` that is wrong / misleading - should be `WrappedSerdeFactory` :D

**Comment by rmoff** [comment]
Thanks @big-andy-coates. We do use ksql-datagen in `cp-demo`, `cp-docker-images`, `examples` (/cc @confluentinc/integration-architecture)  - but if it's backward compatible then I don't think we'll bother to change them. @JimGalasyn can comment from docs point of view if tutorials need updating.

**Pull request merged**